### ISSUE TITLE
Compilation issue "no matching constructor" occurring from time to time.

### ISSUE THREAD

**Issue by rhaidiz** [therecipe/qt#889]
I've been trying to figure this out for the whole day and I've finished the ideas. I couldn't even generate a simple example to demonstrate the issue since I haven't understood what is causing it, so I'll try to describe it.

I'm developing a GUI interface divided in tabs, the main interface is a QTabWidget and each module creates a new tab. 

This is the creation of the main UI, which is part of the core package.
```go
package core

import(
	"github.com/therecipe/qt/widgets"
	"github.com/therecipe/qt/core"
)

type Maingui struct {
	widgets.QMainWindow
	_ func() `constructor:"setup"`
	
	tabWidget			*widgets.QTabWidget
}


func(g *Maingui) setup(){
	
	g.SetWindowTitle("Maingui")
	g.SetMinimumSize(core.NewQSize2(323, 117))

	g.tabWidget = widgets.NewQTabWidget(nil)
	g.tabWidget.SetDocumentMode(true);

	g.SetCentralWidget(g.tabWidget)
}


func(g *Maingui) AddGuiModule(m GuiModule){
	g.tabWidget.AddTab(m.GetModuleGui(), m.Name())
}
```

This is still part of the core package and creates a new session which initialises the main GUI.
```Go
package core

type Session struct{

	Path			string

	// List of modules
	Modules		[]Module

	MainGui		*Maingui
}


func NewSession(path string) *Session{
	return &Session{
		Path: 			path,
		MainGui:		NewMainguii(nil,0),
	}
}

func LoadSession(path string) *Session{
//TODO: Load session from file
return nil
}

func(s *Session)LoadModule(m Module, g GuiModule){
	s.Modules = append(s.Modules, m)
	s.MainGui.AddGuiModule(g)
}
```

Then I have a model that defines a custom table model which is used to create a QTableView for which I've performed very useless changes to the example I found [here](https://github.com/therecipe/examples/blob/master/advanced/widgets/tableview/main.go).

```Go
package model

import(
		"github.com/therecipe/qt/core"
	)
type HItem struct{
	Method			string
}


type CustomTableModel struct {
	core.QAbstractTableModel
	_ func() `constructor:"init"`

	modelData []HItem
}

func (m *CustomTableModel) init() {
	m.modelData = []HItem{{Method: "Test"}}


	m.ConnectHeaderData(m.headerData)
	m.ConnectRowCount(m.rowCount)
	m.ConnectColumnCount(m.columnCount)
	m.ConnectData(m.data)
}

func(m *CustomTableModel) SetData(mp []HItem){
	m.modelData = mp
}

func (m *CustomTableModel) headerData(section int, orientation core.Qt__Orientation, role int) *core.QVariant {
	if role != int(core.Qt__DisplayRole) || orientation == core.Qt__Vertical {
		return m.HeaderDataDefault(section, orientation, role)
	}

	switch section {
	case 0:
		return core.NewQVariant14("Method")
	case 1:
		return core.NewQVariant14("Method")
	}
	return core.NewQVariant()
}

func (m *CustomTableModel) rowCount(*core.QModelIndex) int {
	return len(m.modelData)
}

func (m *CustomTableModel) columnCount(*core.QModelIndex) int {
	return 2
}

func (m *CustomTableModel) data(index *core.QModelIndex, role int) *core.QVariant {
	if role != int(core.Qt__DisplayRole) {
		return core.NewQVariant()
	}

	item := m.modelData[index.Row()]
	switch m.HeaderData(index.Column(), core.Qt__Horizontal, role).ToString() {
	case "Method":
		return core.NewQVariant14(item.Method)
	case "Method":
		return core.NewQVariant14(item.Method)
	}
	return core.NewQVariant()
}
```

The issue I'm having is that, from time to time, I get an error such as this one:

```
moc.cpp:509:19: error: unknown type name 'Maingui'
moc.cpp:510:49: error: unknown type name 'Maingui'
moc.cpp:510:14: error: no matching constructor for initialization of 'CustomTableModel58354e'
moc.cpp:48:7: note: candidate constructor (the implicit copy constructor) not viable: cannot convert argument of incomplete type 'void *' to 'const CustomTableModel58354e' for 1st argument
moc.cpp:52:2: note: candidate constructor not viable: cannot convert argument of incomplete type 'void *' to 'QObject *' for 1st argument
```

What I cannot figure out is why this happens only from time to time. I have the feeling that there's come sort of concurrent (maybe?) issue while compiling, and that one time it does not generate the constructor for the CustomTableModel while a second or third time does create the constructor so everything is ok

I really have no idea if I'm doing something wrong or not. Any suggestion?

**EDIT:** Further information, I compared moc.cpp after a *successful* compilation with a moc.cpp after an *unsuccessful* compilation. Here's the diff:

Successful
```cpp
void* CustomTableModel58354e_NewCustomTableModel(void* parent)
{
	if (dynamic_cast<QCameraImageCapture*>(static_cast<QObject*>(parent))) {
		return new CustomTableModel58354e(static_cast<QCameraImageCapture*>(parent));
```

Unsuccessful (notice the `Maingui` thing)
```cpp
void* CustomTableModel58354e_NewCustomTableModel(void* parent)
{
	if (dynamic_cast<Maingui*>(static_cast<QObject*>(parent))) {
		return new CustomTableModel58354e(static_cast<Maingui*>(parent));
```

What am I missing??

**Comment by therecipe** [comment]
Hey

Which version of the binding do you use? (you can get the commit with `$GOPATH/bin/qtsetup check`)
What class does `Broxygui` inherit?
Did you try to remove all `moc*` files and then re-ran qtdeploy (or `qtmoc && go build`?)

And does this example work for you?
```
github.com/therecipe/abc
├── core
│   ├── maingui.go
│   └── session.go
├── main.go
└── model
    └── model.go
```

`core/maingui.go`
```go
package core

import (
	"github.com/therecipe/qt/core"
	"github.com/therecipe/qt/widgets"

	. "github.com/therecipe/abc/model"
)

type Maingui struct {
	widgets.QMainWindow
	_ func() `constructor:"setup"`

	tabWidget *widgets.QTabWidget
}

func (g *Maingui) setup() {

	g.SetWindowTitle("Maingui")
	g.SetMinimumSize(core.NewQSize2(323, 117))

	g.tabWidget = widgets.NewQTabWidget(nil)
	g.tabWidget.SetDocumentMode(true)

	g.SetCentralWidget(g.tabWidget)
}

func (g *Maingui) AddGuiModule(m *CustomTableModel) {
	g.tabWidget.AddTab(m.GetModuleGui(), m.Name())
}
```
`core/session.go`
```go
package core

import . "github.com/therecipe/abc/model"

type Session struct {
	Path string

	// List of modules
	Modules []*CustomTableModel

	MainGui *Maingui
}

func NewSession(path string) *Session {
	return &Session{
		Path:    path,
		MainGui: NewMaingui(nil, 0),
	}
}

func LoadSession(path string) *Session {
	//TODO: Load session from file
	return nil
}

func (s *Session) LoadModule(m *CustomTableModel, g *CustomTableModel) {
	s.Modules = append(s.Modules, m)
	s.MainGui.AddGuiModule(g)
}
```
`model/model.go`
```go
package model

import (
	"github.com/therecipe/qt/core"
	"github.com/therecipe/qt/widgets"
)

type HItem struct {
	Method string
}

type CustomTableModel struct {
	core.QAbstractTableModel
	_ func() `constructor:"init"`

	modelData []HItem
}

func (m *CustomTableModel) Name() string {
	return "someName"
}

func (m *CustomTableModel) GetModuleGui() *widgets.QTableView {
	v := widgets.NewQTableView(nil)
	v.SetModel(m)
	return v
}

func (m *CustomTableModel) init() {
	m.modelData = []HItem{{Method: "Test"}}

	m.ConnectHeaderData(m.headerData)
	m.ConnectRowCount(m.rowCount)
	m.ConnectColumnCount(m.columnCount)
	m.ConnectData(m.data)
}

func (m *CustomTableModel) SetData(mp []HItem) {
	m.modelData = mp
}

func (m *CustomTableModel) headerData(section int, orientation core.Qt__Orientation, role int) *core.QVariant {
	if role != int(core.Qt__DisplayRole) || orientation == core.Qt__Vertical {
		return m.HeaderDataDefault(section, orientation, role)
	}

	switch section {
	case 0:
		return core.NewQVariant1("Method")
	case 1:
		return core.NewQVariant1("Method")
	}
	return core.NewQVariant()
}

func (m *CustomTableModel) rowCount(*core.QModelIndex) int {
	return len(m.modelData)
}

func (m *CustomTableModel) columnCount(*core.QModelIndex) int {
	return 2
}

func (m *CustomTableModel) data(index *core.QModelIndex, role int) *core.QVariant {
	if role != int(core.Qt__DisplayRole) {
		return core.NewQVariant()
	}

	item := m.modelData[index.Row()]
	switch m.HeaderData(index.Column(), core.Qt__Horizontal, role).ToString() {
	case "Method":
		return core.NewQVariant1(item.Method)
	}
	return core.NewQVariant()
}
```
`main.go`
```go
package main

import (
	"os"

	"github.com/therecipe/qt/widgets"

	"github.com/therecipe/abc/core"
	"github.com/therecipe/abc/model"
)

func main() {
	widgets.NewQApplication(len(os.Args), os.Args)

	mod := model.NewCustomTableModel(nil)
	s := core.NewSession("")
	s.LoadModule(mod, mod)
	s.MainGui.Show()

	widgets.QApplication_Exec()
}
```

**Comment by rhaidiz** [comment]
Sorry, `Broxygui` was a copy\paste mistake. It was the old name of `Maingui` which I renamed for simplicity and my clipboard mixed the paste (I fixed my first comment).

This is the output of `qtsetup check`:

```./qtsetup check                                                                                                                                                                                                                            Mon Jun 24 22:24:34 2019
INFO[0000] running: 'qtsetup check darwin' [docker=false] [vagrant=false]
INFO[0000] GOOS:                        'darwin'
INFO[0000] GOARCH:                      'amd64'
INFO[0000] GOVERSION:                   'go1.11.5'
INFO[0000] GOROOT:                      '/usr/local/go'
INFO[0000] GOPATH:                      '/Users/rhaidiz/go'
INFO[0000] GOBIN:                       '/Users/rhaidiz/go/bin'
INFO[0000] GOMOD:                       ''
INFO[0000] QT_HASH:                     'cf6956137ad2a2b502ae3cf8b4f8026bec87364d'
INFO[0000] QT_API:                      ''
INFO[0000] QT_VERSION:                  '5.12.0'
INFO[0000] QT_DIR:                      '/Users/rhaidiz/Qt'
INFO[0000] QT_STUB:                     'false'
INFO[0000] QT_DEBUG:                    'false'
INFO[0000] QT_QMAKE_DIR:                ''
INFO[0000] QT_WEBKIT:                   'false'
INFO[0000] QT_HOMEBREW:                 'false'
INFO[0000] QT_MACPORTS:                 'false'
INFO[0000] QT_NIX:                      'false'
INFO[0000] QT_FELGO:                    'false'
INFO[0000] XCODE_DIR:                   '/Applications/Xcode.app'
```

Your example does not compile but gives a different error:

```
github.com/rhaidiz/abc/model
# github.com/rhaidiz/abc/model
model/model.go:49:10: undefined: core.NewQVariant1
model/model.go:51:10: undefined: core.NewQVariant1
model/model.go:72:10: undefined: core.NewQVariant1
```

**Comment by therecipe** [comment]
No problem, about the `core.NewQVariant1` you will probably just need to replace it with `core.NewQVariant14` for it to work. (you seem to be 2 commits behind the commit that added the function)

Does your `model` package import your `core` package? Or just the `qt/core` package?

Also could you try to `go fmt ./...` your code? I looked into this and it should theoretically
not be possible for `Maingui` to show up in the `NewCustomTableModel`, but I can recall some issue with unformatted go code which was similar to your issue I think.

**Comment by rhaidiz** [comment]
Yeah I figured the `NewQVariant1` later 😅 

My `model` does not import my `core`, it only imports the `qt/core` package.

I removed all the `moc*` files and did a `go fmt ./..` as you suggested and it seems to be working now. I also tried to perform a couple of changes here and there on my `model` (like adding an extra column) and still seems to be working just fine.

If it's ok by you, before closing the issue, I'll keep doing some more editing to make sure that keeping the go code properly formatted actually fixes the issue. It seems it does fix it, but since it doesn't always happen I want to make sure.

For know, thank you so much for your help 😃
